torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -588,6 +656,127 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce_scatter(
+    checkSplitSizes(inputSplitSizes, inputTensor, size_);
+    checkSplitSizes(outputSplitSizes, outputTensor, size_);
+    std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this, inputSplitSizes, outputSplitSizes](std::unique_ptr<WorkEntry>& entry) {
+        auto srcdata = (entry->src)[0];
+        auto dstdata = (entry->dst)[0];
+        std::vector<int> send_lengths(size_);
+        std::vector<int> recv_lengths(size_);
+        std::vector<int> send_offsets(size_);
+        std::vector<int> recv_offsets(size_);
+        computeLengthsAndOffsets(inputSplitSizes, srcdata, send_lengths, send_offsets, size_);
+        computeLengthsAndOffsets(outputSplitSizes, dstdata, recv_lengths, recv_offsets, size_);
+        c10::DeviceGuard guard(srcdata.device());
+        std::unique_lock<std::mutex> globalLock(pgGlobalMutex_);
+        MPI_CHECK(MPI_Alltoallv(
+              srcdata.data_ptr(),
+              send_lengths.data(),
+              send_offsets.data(),
+              mpiDatatype.at(srcdata.scalar_type()),
+              dstdata.data_ptr(),
+              recv_lengths.data(),
+              recv_offsets.data(),
+              mpiDatatype.at(dstdata.scalar_type()),
+              pgComm_));
+      };
+    std::vector<at::Tensor> inputTensors = {inputTensor};
+    std::vector<at::Tensor> outputTensors = {outputTensor};
+    auto entry = std::unique_ptr<WorkEntry>(
+        new WorkEntry(&inputTensors, &outputTensors, std::move(runFunc)));
+    return enqueue(std::move(entry));
+  }
+}
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::alltoall(
+    std::vector<at::Tensor>& outputTensors,
+    std::vector<at::Tensor>& inputTensors,
+    const AllToAllOptions& opts) {
+  if (inputTensors.size() != size_ || outputTensors.size() != size_) {
+    throw std::runtime_error("Number of input or output tensors are not equal to group size");
+  }
+  std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+    [opts, this](std::unique_ptr<WorkEntry>& entry) {
+      std::vector<int> send_lengths(size_);
+      std::vector<int> recv_lengths(size_);
+      std::vector<int> send_offsets(size_);
+      std::vector<int> recv_offsets(size_);
+      auto srcdata = entry->src;
+      auto dstdata = entry->dst;
+      int64_t src_len = computeLengthsAndOffsets(srcdata, send_lengths, send_offsets, size_);
+      int64_t dst_len = computeLengthsAndOffsets(dstdata, recv_lengths, recv_offsets, size_);
+      std::vector<int64_t> send_lengthsL(send_lengths.begin(), send_lengths.end());
+      std::vector<int64_t> recv_lengthsL(recv_lengths.begin(), recv_lengths.end());
+      at::Tensor srcFlatData = at::empty({src_len}, srcdata[0].options());
+      at::Tensor dstFlatData = at::empty({dst_len}, dstdata[0].options());
+      auto srcFlatDataSplits = srcFlatData.split_with_sizes(c10::IntArrayRef(send_lengthsL), 0);
+      for (int i = 0; i < size_; i++) {
+        srcFlatDataSplits[i].copy_(srcdata[i].view({-1}));
+      }
```

Comment:
let's use primitives that other process groups are already using, e.g. flattenDenseTensors and newLikeFlat?

Reply:
Sorry, but those primitives won't work here as they need equal sized tensors and we may have unequal sized tensors in the list here.

Reply:
you are right, apologies! I think flattenDenseTensors should still work for your usecase (there is no limitation on sizes there)? newLikeFlat won't work unfortunately...

Reply:
For now I am keeping it as is, but plan to optimize it later to check if list of tensors is derived from flat tensor. Will change this at that time.